42.md

```diff
@@ -56,11 +56,23 @@ Relays MUST exclude `kind: 22242` events from being broadcasted to any client.
 
 ## Protocol flow
 
-At any moment the relay may send an `AUTH` message to the client containing a challenge. After receiving that the client may decide to
-authenticate itself or not. The challenge is expected to be valid for the duration of the connection or until a next challenge is sent by
-the relay.
+At specific moments the relay may send an `AUTH` message to the client containing a challenge, in one of the following forms:
 
-The client may send an auth message right before performing an action for which it knows authentication will be required -- for example, right
+- `["AUTH", <challenge-string>]`: No third array element. It means relay requires authentication for any resource access.
+If applicable, it MUST be sent right after WebSocket `connection`.
+Client may have already sent a message before receiving this one, in that case relay MUST also send an `AUTH` message in one of the other forms below;
+- `["AUTH", <challenge-string>, "sub:<subscription_id>"]`: It references a previous client `REQ` message with a specific subscription id as third array element.
+It means client subscription requested atleast one event that requires prior authentication to be accessed.
+MUST be sent before the `EOSE` message with same subscription id. The relay SHOULD send no events before `EOSE` because the client is expected to retry the `REQ` after authentication;
+- `["AUTH", <challenge-string>, "evt:<event_id>"]`: It references a previous client `EVENT` message with a specific event id as third array element.
+It means relay requires authentication before publishing any or atleast this kind of event.
+MUST be sent before `OK` message with same event id.
+
+*Note: The challenge string is the same during the whole WebSocket connection. Multiple relay `AUTH` messages on the same connection
+carry the same challenge string.*
```

Comment:
I admit this is theoretical, but I think it could be useful to authenticate multiple times during a session without closing the connection and re-opening it (although that's probably a fine approach too). What about this:

> The challenge string SHOULD be the same for the duration of the client session. Relays MAY change the challenge string if they wish to indicate that a new session with different authentication requirements has been initiated (for example when changing identities or making a claim with a separate identity).

For now, it's probably best to just keep this open-ended and say:

> The challenge string SHOULD be the same for the duration of the client session. 

Separately, it would be good to clarify that the relay can send multiple AUTH messages specifying different messages the client attempted to send:

> Relays SHOULD issue the same challenge multiple times, each time indicating a different message the client has sent that was not accepted due to AUTH.

Reply:
> ...for example when changing identities...

I think this was ditched in the past in favor of simply reconnecting. For example, #891 shouldn't be used as a way of switching accounts.

I prefer lowering complexity but I will change it to your version as a middle ground.

Reply:
> I prefer lowering complexity but I will change it to your version as a middle ground.

No, I think we should leave this out for now. I think we'll want it eventually, but it's better to hew to YAGNI when possible. Re-connecting is a fine way to solve this in the near term.

Reply:
Ok didn't change it